dashboard/lib/controllers/batchQueue/index.js

```diff
@@ -0,0 +1,38 @@
+class BatchQueue {
+    constructor(queueProcessor) {
+        this.queue = [];
+        this.processing = false;
+        this.queueProcessor = queueProcessor;
+    }
+
+    add(data) {
+        this.queue.push(data);
+        this.processQueue();
+    }
+
+    async processQueue() {
+        if (this.queue.length == 0) {
+            this.processing = false;
+            return;
+        }
+
+        if (this.processing) return;
+
+        this.processing = true;
+        const currentQueue = this.queue;
+        this.queue = [];
+
+        await currentQueue.reduce(async (prevPromise, item, index) => {
+            await prevPromise;
+            await this.queueProcessor(item);
+
+            // at last, mark processing as complete so that new batch is saved
+            if (index == currentQueue.length - 1) {
+                this.processing = false;
```

Comment:
Let it be handled on line 15. Just call this method after the `reduce`. It'll take care of processing the next batch. It'll keep our loop clean.

Reply:
Oh okay, I also tried that initially but thought it would lead to too many calls. Would this be a decent tradeoff?

Reply:
It would be just 1 extra call, no? The benefit would be that changing the flag would be kept at a single place making it easier to read and track.